presto-client/src/main/java/io/prestosql/client/JavaVersion.java

```diff
@@ -0,0 +1,138 @@
+
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.client;
+
+import com.google.common.base.StandardSystemProperty;
+
+import java.util.Objects;
+import java.util.Optional;
+import java.util.OptionalInt;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static java.lang.String.format;
+
+// TODO: remove this when we upgrade to Java 9 (replace with java.lang.Runtime.getVersion())
+//       It's copied from io.prestosql.server module.
+public class JavaVersion
```

Comment:
I’m concerned about having this complex code (which has had bugs in the past) as part of the JDBC driver, which runs in various environments. We also plan to keep the driver on Java 8 for the foreseeable future, so this code won’t be removed anytime soon.

Some alternatives:

1. Just split on a period and take the first item, since that works for all versions.

2. Use reflection to check for and call the `Runtime.Version` API. We can check for and use the `getFeature` method which is in JDK 10+.

3. Use reflection to look for the existence of a “marker” class like `java.net.http.HttpClient` which was added in JDK 11.

Reply:
This is the code using `Runtime.version()` reflectively:
```java
@SuppressWarnings("JavaReflectionMemberAccess")
private static boolean isAtLeastJava11()
{
    try {
        Method versionMethod = Runtime.class.getMethod("version");
        Method featureMethod = versionMethod.getReturnType().getMethod("feature");
        Object version = versionMethod.invoke(null);
        int feature = (int) featureMethod.invoke(version);
        return feature >= 11;
    }
    catch (ReflectiveOperationException e) {
        return false;
    }
}

Reply:
Code to safely parse the version:
```java
private static boolean isAtLeastJava11()
{
    String feature = Splitter.on(".").split(StandardSystemProperty.JAVA_VERSION.value()).next();
    try {
        return Integer.parseInt(feature) >= 11;
    }
    catch (NumberFormatException e) {
        return false;
    }
}

Reply:
@dain agrees we should go with option 1, since that's simple and reliable